apps/genrsa.c

```diff
@@ -168,49 +165,78 @@ int genrsa_main(int argc, char **argv)
     if (out == NULL)
         goto end;
 
+    if (!init_gen_str(&ctx, "RSA", eng, 0))
+        goto end;
+
+    EVP_PKEY_CTX_set_cb(ctx, genrsa_cb);
+    EVP_PKEY_CTX_set_app_data(ctx, bio_err);
+
+    if (EVP_PKEY_CTX_set_rsa_keygen_bits(ctx, num) <= 0) {
+        BIO_printf(bio_err, "Error setting RSA length\n");
+        goto end;
+    }
+    if (!BN_set_word(bn, f4)) {
+        BIO_printf(bio_err, "Error allocating RSA public exponent\n");
+        goto end;
+    }
+    if (EVP_PKEY_CTX_set_rsa_keygen_pubexp(ctx, bn) <= 0) {
+        BIO_printf(bio_err, "Error setting RSA public exponent\n");
+        goto end;
+    }
+    if (EVP_PKEY_CTX_set_rsa_keygen_primes(ctx, primes) <= 0) {
+        BIO_printf(bio_err, "Error setting number of primes\n");
+        goto end;
+    }
     if (verbose)
         BIO_printf(bio_err, "Generating RSA private key, %d bit long modulus (%d primes)\n",
                    num, primes);
-    rsa = eng ? RSA_new_method(eng) : RSA_new();
-    if (rsa == NULL)
-        goto end;
-
-    if (!BN_set_word(bn, f4)
-        || !RSA_generate_multi_prime_key(rsa, num, primes, bn, cb))
+    if (!EVP_PKEY_keygen(ctx, &pkey)) {
+        BIO_printf(bio_err, "Error generating RSA key\n");
         goto end;
+    }
 
-    RSA_get0_key(rsa, NULL, &e, NULL);
-    hexe = BN_bn2hex(e);
-    dece = BN_bn2dec(e);
-    if (hexe && dece && verbose) {
-        BIO_printf(bio_err, "e is %s (0x%s)\n", dece, hexe);
+    if (verbose) {
+        ebuf = app_malloc(num / 8, "e buffer");
+        *params = OSSL_PARAM_construct_BN("e", ebuf, num / 8);
+        if (!EVP_PKEY_CTX_set_params(ctx, params)) {
```

Comment:
You probably mean `EVP_PKEY_CTX_get_params`

Reply:
I don't think so.  I want to get the _e_ value out.  There doesn't seem to be a call to get a GEN-param, they can only be set.

Reply:
Er, you want to *get* `e`'s value, right? Not set it?